modules/apps/redirect/redirect-service/src/main/java/com/liferay/redirect/internal/upgrade/v3_0_2/RedirectEntrySourceURLUpgradeProcess.java

```diff
@@ -0,0 +1,121 @@
+ */
+
+package com.liferay.redirect.internal.upgrade.v3_0_2;
+
+import com.liferay.petra.string.StringBundler;
+import com.liferay.portal.kernel.dao.jdbc.AutoBatchPreparedStatementUtil;
+import com.liferay.portal.kernel.log.Log;
+import com.liferay.portal.kernel.log.LogFactoryUtil;
+import com.liferay.portal.kernel.upgrade.UpgradeProcess;
+import com.liferay.portal.kernel.util.StringUtil;
+
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * @author Attila Bakay
+ */
+public class RedirectEntrySourceURLUpgradeProcess extends UpgradeProcess {
+
+	@Override
+	protected void doUpgrade() throws Exception {
+		try (PreparedStatement preparedStatement1 = connection.prepareStatement(
+				"select redirectEntryId, sourceURL from RedirectEntry");
+			PreparedStatement preparedStatement2 =
+				AutoBatchPreparedStatementUtil.concurrentAutoBatch(
+					connection,
+					"update RedirectEntry set sourceURL = ? where " +
+						"redirectEntryId = ?")) {
+
+			ResultSet resultSet = preparedStatement1.executeQuery();
+
+			while (resultSet.next()) {
+				String sourceURL = resultSet.getString(2);
+				long redirectEntryId = resultSet.getLong(1);
+
+				_redirectEntries.put(redirectEntryId, sourceURL);
+			}
+
+			for (Map.Entry<Long, String> redirectEntry :
+					_redirectEntries.entrySet()) {
+
+				String sourceURL = redirectEntry.getValue();
+
+				String lowerCaseSourceURL = StringUtil.toLowerCase(sourceURL);
+
+				if (sourceURL.equals(lowerCaseSourceURL)) {
+					continue;
+				}
+
+				long redirectEntryId = redirectEntry.getKey();
+
+				if (!_hasRedirectEntry(redirectEntryId, lowerCaseSourceURL)) {
+					preparedStatement2.setString(1, lowerCaseSourceURL);
+					preparedStatement2.setLong(2, redirectEntryId);
+
+					preparedStatement2.addBatch();
+				}
+				else {
+					if (_log.isWarnEnabled()) {
+						_log.warn(
+							StringBundler.concat(
+								"Can not modify '", sourceURL, "' to '",
+								lowerCaseSourceURL,
+								"' because it is already in use id:  at id: ",
+								redirectEntryId));
+					}
+				}
+			}
+
+			preparedStatement2.executeBatch();
+		}
+	}
+
+	private boolean _hasRedirectEntry(
+		long redirectEntryId, String lowerCaseSourceURL) {
+
+		if (_duplicateSourceURLs.contains(lowerCaseSourceURL)) {
+			return true;
+		}
+
+		_duplicateSourceURLs.add(lowerCaseSourceURL);
+
+		for (Map.Entry<Long, String> redirectEntry :
+				_redirectEntries.entrySet()) {
+
+			String sourceURL = redirectEntry.getValue();
+
+			if ((redirectEntry.getKey() != redirectEntryId) &&
+				sourceURL.equals(lowerCaseSourceURL)) {
+
+				return true;
+			}
+		}
+
+		return false;
+	}
+
+	private static final Log _log = LogFactoryUtil.getLog(
+		RedirectEntrySourceURLUpgradeProcess.class);
+
+	private final Set<String> _duplicateSourceURLs = new HashSet<>();
+	private final Map<Long, String> _redirectEntries = new HashMap<>();
```

Comment:
Can you use a LinkedHashMap so that the iteration order is the same as insertion order?

Reply:
I fixed this one.